pkg/controller/operators/catalog/subscription/reconciler.go

```diff
@@ -109,6 +122,105 @@ func (c *catalogHealthReconciler) Reconcile(ctx context.Context, in kubestate.St
 	return
 }
 
+// updateDeprecatedStatus adds deprecation status conditions to the subscription when present in the cache entry
+func (c *catalogHealthReconciler) updateDeprecatedStatus(ctx context.Context, sub *v1alpha1.Subscription) (bool, error) {
+	if c.sourceProvider == nil {
+		return false, nil
+	}
+
+	source, ok := c.sourceProvider.Sources(sub.Namespace)[cache.SourceKey{
+		Name:      sub.Spec.CatalogSource,
+		Namespace: sub.Namespace,
+	}]
+	if !ok {
+		return false, nil
+	}
+	snapshot, err := source.Snapshot(ctx)
+	// return err if it isn't nil or err (which is nil) if the entries is an empty list.
+	if err != nil || len(snapshot.Entries) == 0 {
+		return false, err
+	}
+
+	changed := false
+	rollupMessages := []string{}
+	var deprecations *cache.Deprecations
+	for _, entry := range snapshot.Entries {
+		// Find the cache entry that matches this subscription
+		if entry.SourceInfo == nil || entry.Package() != sub.Spec.Package {
+			continue
+		}
+		deprecations = &entry.SourceInfo.Deprecations
+		break
+	}
+	if deprecations == nil {
+		// No matching entry found
+		return false, nil
+	}
```

Comment:
In all of the cases that we `return false, nil` here and up, that means that whatever the subscription currently says is what we continue going with, right?

I wonder if we need to specifically set the subscription deprecation conditions to `Unknown` in some (or all?) of these cases?

Reply:
@everettraven Yes I think you're correct regarding Joe's meaning; however i don't think that setting these conditions to `Unknown` is necessary. It adds additional complexity to this logic which is already getting somewhat complex, and additionally in my mind it's not even necessarily correct.

If an installed bundle is deprecated, then it will be marked as such on the cluster. If that bundle is then removed from the catalog, isn't the bundle still deprecated? The only situation I can think of where this might be useful is a package/channel/bundle is marked deprecated in a catalog, removed from the catalog, but then they decide later that it should not have been deprecated, but do not put it back in the catalog.

It would also require some refactoring because the Subscription Status' `GetCondition` method relies on the `Unknown` Condition to work correctly (although it shouldn't because of this kind of situation).

If I'm misunderstanding here definitely let me know though!

Reply:
Yes, sorry I was suggesting using `status: Unknown` on the existing types, not suggesting a new types.

I'm actually curious what OLMv0 does with subscriptions that have been successfully installed, but then later have those that subscription's package, channel, or bundle removed from the catalog. Do we surface that information somehow? I assume at a minimum we'd somehow say, "couldn't resolve this subscription anymore".

But regardless, I'm also concerned about having "sticky" state that depends on what happened in the past. Consider two users (A and B) and three versions of a catalog:
1. Has bundle foo.v1 with no deprecation
2. Has bundle foo.v1 with a deprecation
3. Removes bundle foo.v1 entirely

If user A sees version 1, 2, and 3 of the catalog, their subscription will be stuck on "foo.v1 is deprecated"
If user B sees only versions 1 and 3 of the catalog, their subscription will be stuck on  "foo.v1 is not deprecated"

That strikes me as (a) undesirable in general and (b) totally possible, especially with disconnected clusters which often pick up catalog changes on some infrequent basis and do heavy catalog pruning during the mirroring step to minimize time/size of the transfer.

Reply:
I agree with @dtfranz here about the unknown status being ripe for confusion and solving the "sticky" state as a follow up.

For me this raises the question of whether or not we have thought of best practices/recommendations for catalog maintainers to follow? It strikes me as odd that we are attempting to solve for the many edge cases around how someone might create and modify their catalog and the versioning (after all it _is_ an OCI image and we don't have any control over how users make changes to it) instead of coming up with a set of guidelines for how OLM interacts with the catalogs and what "expectations" are in place. This way we can note that if a catalog maintainer deviates from the recommendations/best practices things may not function as expected due to the assumptions/expectations that have been made in OLM to provide the best user experience possible. No need to have this conversation in this PR, but something I would definitely be interested in having a continued discussion about as I think it also affects the OLMv1 work.

Reply:
I'm suppose I'm okay with existing behavior since my scenario is an edge case (though not a total edge case because it _is_ expected for a deprecated bundle to eventually be removed from the catalog)

Can we capture a follow-up issue that captures both of the following?
- what we do generally to inform users that their currently installed stuff is no longer present in the catalog.
- specifically, what the implications are for the new deprecation condition types.

Reply:
Thank you @joelanford! I've created issue #3122 to continue this discussion.

Reply:
>I'm actually curious what OLMv0 does with subscriptions that have been successfully installed, but then later have thier package, channel, or bundle removed from the catalog. Do we surface that information somehow? I assume at a minimum we'd somehow say, "couldn't resolve this subscription anymore".

@joelanford I believe that the following is correct, but I'll rely on @kevinrizza to keep me honest:

- in the case of a package or channel removal, the resolution failure is surfaced in the subscription.
- the removed bundle case is more nuanced. Resolution will succeed if the version falls within the skip-range of a separate bundle. The bundle shouldn't have been removed if replaces or skips was used, so it'd be the case of an invalid catalog, the resolution would fail and be surfaced in the subscription.

>But regardless, I'm also concerned about having "sticky" state that depends on what happened in the past. >Consider two users (A and B) and three versions of a catalog:

    > Has bundle foo.v1 with no deprecation
    > Has bundle foo.v1 with a deprecation
    > Removes bundle foo.v1 entirely

>If user A sees version 1, 2, and 3 of the catalog, their subscription will be stuck on "foo.v1 is deprecated"
>If user B sees only versions 1 and 3 of the catalog, their subscription will be stuck on "foo.v1 is not deprecated"

>That strikes me as (a) undesirable in general and (b) totally possible, especially with disconnected clusters which often pick up catalog changes on some infrequent basis and do heavy catalog pruning during the mirroring step to minimize time/size of the transfer.

Agreed, worth tracking in an issue as suggested in this thread.